flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.adaptive.allocator;
+
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.jobmanager.scheduler.SlotSharingGroup;
+import org.apache.flink.runtime.scheduler.TaskBalancedExecutionSlotSharingGroupBuilder;
+import org.apache.flink.runtime.scheduler.adaptive.allocator.SlotSharingSlotAllocator.ExecutionSlotSharingGroup;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * The implementation of the {@link SlotSharingStrategy} based on tasks balanced strategy. TODO: The
+ * implementation will be used in the followed jira issues done.
+ * https://issues.apache.org/jira/browse/FLINK-35966
+ * https://issues.apache.org/jira/browse/FLINK-33391.
+ */
+enum TaskBalancedSlotSharingStrategy implements SlotSharingStrategy {
+    INSTANCE;
+
+    @Override
+    public List<ExecutionSlotSharingGroup> getExecutionSlotSharingGroups(
+            JobInformation jobInformation, VertexParallelism vertexParallelism) {
+        return new TaskBalancedExecutionSlotSharingGroupBuilder(
+                        getAllVertices(jobInformation, vertexParallelism),
+                        jobInformation.getSlotSharingGroups(),
+                        jobInformation.getCoLocationGroups())
+                .build().values().stream()
+                        .map(
+                                fromGroup ->
+                                        new ExecutionSlotSharingGroup(
+                                                fromGroup.getExecutionVertexIds()))
+                        .collect(Collectors.toList());
```

Comment:
```suggestion
            .build().values().stream()
                        .distinct()
                        .map(
                                fromGroup ->
                                        new ExecutionSlotSharingGroup(
                                                fromGroup.getExecutionVertexIds()))
                        .collect(Collectors.toList());
```
missing **distinct**

Reply:
thx for the catching~. 👍 
I'll add a related test case for it.

Reply:
Add the test cases in TaskBalancedSlotSharingStrategyTest